### PR TITLE
Oceanwater 750 improve identify sample

### PR DESCRIPTION
**Pull request by Keeganfn**
### Overview:
Improved the IdentifySampleTarget plan. Plan now uses the Stereo Camera to take pictures of the ground in front of the lander and processes them in a separate action server using OpenCV. User can find points of interest using 1 of 2 filters "Dark" and "Brown" (finding dark and brown spots respectively). The points of interest are then converted to 3d points using the point cloud and the best suited point is sent back as the result to the OwAdapter and OwInterface classes for use in the plan.

### Testing:
1. ```roslaunch ow atacama_y1a.launch```
2. ```roslaunch ow_plexil plexil_node.launch``` (Now also launches the identify_sample_location node)
3. Run either IdentifySampleLocationDemo.plx (shows both filters working) or ReferenceMission1.plx (Shows working in real mission)
#### Optional Visualization:
4. In Rviz to see the 3d point visualized add a Marker and connect to the ```/sample_point_visualization``` topic
5. In the image viewer the ```/sample_location``` topic will show you the selected sample point and the contours.

If a point is outputted to the terminal and Guarded Move begins successfully (Or no point is found and exits gracefully) it is working!

### Notes:
1. Also fixed the Camera subscriber not waiting for the StereoCamera/points2 topic to be published before marking TakePicture as finished in OwInterface. This caused some point clouds not to be published if pictures were taken in succession. Terry recommended adding a timeout which is implemented but a bit clunky, if you have any suggestions to make it more streamlined Id love to hear it!

2. All image processing is done in the IdentifySampleTarget python class in the Scripts directory before sending a single point back as a result. 

3. Sometimes a point close to the lander will cause the guarded move to fail. I don't think this is a problem on my end as it looks to be within the range of motion of the arm, but I could be wrong. Unpredictable when a point will cause this issue.

4. The brown filter seems a little more realistic but due to the subtly of color within the environment the Dark filter sometimes performs better.

Thanks!

### CODE REVIEW

**Comment by kmdalal** [comment]
My first run resulted in this error:
```
Enter any additional plan to be run (or use the GUI): 
IdentifySampleLocationDemo.plx
[ INFO]: Starting TiltAntenna: 50.000000 degrees (0.872665 radians)
[ INFO]: Starting PanAntenna: 65.000000 degrees (1.134464 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Starting PanAntenna: 70.000000 degrees (1.221730 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Starting PanAntenna: 75.000000 degrees (1.308997 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Sending goal to action IdentifySampleLocation
[ INFO]: IdentifySampleLocation started...
[ERROR]: Exception in your execute callback: too many values to unpack
Traceback (most recent call last):
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/simple_action_server.py", line 289, in executeLoop
    self.execute_callback(goal)
  File "/home/kdalal/ow/src/ow_autonomy/ow_plexil/scripts/identify_sample_location.py", line 74, in sample_location_callback
    sample_points_2d, contour_areas, cv_image = self.identify_sample_spots(i[0], goal.filter_type)
  File "/home/kdalal/ow/src/ow_autonomy/ow_plexil/scripts/identify_sample_location.py", line 147, in identify_sample_spots
    contours, hierarchy = cv.findContours(thresh, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
ValueError: too many values to unpack

[ERROR]: Could not get sample point from IdentifySampleLocation
```

**Comment by kmdalal** [comment]
I'm not certain I've installed the correct version of OpenCV.   The given 'pip' command didn't work for me (errors) and so I used 'apt get'.  I accidentally installed the Python 3 version of OpenCV.  Then I tried uninstalling this and reinstalling the Python 2 version.  Perhaps my environment is messed up somehow, but I'm not sure what to check.

**Comment by Keeganfn** [comment]
> I'm not certain I've installed the correct version of OpenCV. The given 'pip' command didn't work for me (errors) and so I used 'apt get'. I accidentally installed the Python 3 version of OpenCV. Then I tried uninstalling this and reinstalling the Python 2 version. Perhaps my environment is messed up somehow, but I'm not sure what to check.

I think it is probably an issue with your install. I looked into it and it seems python 2.7 support is no longer supported in new releases, so you will need to specify the most recent release that supports 2.7. I recreated the install on my system and the command:

```pip install opencv-python==4.2.0.32```

Worked correctly for me, hopefully that will also fix your issues. Once the transition is made to Noetic and Python 3 we wont have to specify the version number anymore.

**Comment by Samahu** [comment]
I ran into the same issue about **python-opencv**. The issue is ROS-melodic ships with OpenCV 3.0. Why do we need install opencv-python 4.0? All is needed to get around the error is to add a placeholder for one extra return value as follows:
```python
[im2], contours, hierarchy = cv.findContours(thresh, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
```

**Comment by Keeganfn** [comment]
Thats good to know, I didn't know that OpenCV 3.0 ships with ROS-Melodic. I have OpenCV 4.0 installed on my machine which might be why I was not recieving an error but you were. I reverted back to OpenCV 3.0 on my machine and was able to recreate the error and fix it with an extra return value like you said. Thanks!

File: ow_plexil/src/plans/IdentifySampleLocationDemo.plp [hunk 1]
```diff
@@ -0,0 +1,62 @@
+// The Notices and Disclaimers for Ocean Worlds Autonomy Testbed for Exploration
```
Comment (Samahu) [inline ow_plexil/src/plans/IdentifySampleLocationDemo.plp:1]: Why wouldn't The `IdentifySampleLocationDemo.plp` use the `IdentifySampleTarget.plp`?

File: ow_plexil/src/plans/IdentifySampleTarget.plp [hunk 2]
```diff
@@ -2,31 +2,47 @@
...
-  // direction are stubbed here; these choices are arbitrary.
-  X = 1.65;
-  Y = 0;
-  Parallel = true;
+  // Take pictures in front of the lander for processing in identify_sample_location
+  // An alternative would be to create a new takepicture command that waits for the pointcloud or modifying the current one
+  LibraryCall Tilt(Degrees = 50);
+  LibraryCall PanAndShoot(PanAngle = 65);
+  LibraryCall PanAndShoot(PanAngle = 70);
+  LibraryCall PanAndShoot(PanAngle = 75);
```
Comment (Samahu) [inline ow_plexil/src/plans/IdentifySampleTarget.plp:26]: I would recommend you define the center position (expressed in angles) and then use +/- for the extra shots. i.e:
```suggestion
  c_tilt = 50;
  c_pan = 70;
  delta = 5;
  LibraryCall Tilt(Degrees = c_tilt);
  LibraryCall PanAndShoot(PanAngle = c_pan - delta);
  LibraryCall PanAndShoot(PanAngle = c_pan);
  LibraryCall PanAndShoot(PanAngle = c_pan + delta);
```

**Review by Samahu — CHANGES_REQUESTED** [review]
I haven't fully gone through your code ... but I made some comments that are worth checking.

- [x] Use geometry_msgs::Point
- [ ] Make `IdentifySampleTargetLocationDemo.plp` reuse `IdentifySampleTarget.plp` 
- [ ] Formalize the PanAndShoot procedure

File: ow_plexil/src/plans/IdentifySampleLocationDemo.plp (hunk 1 above)
Comment (Keeganfn) [reply ow_plexil/src/plans/IdentifySampleLocationDemo.plp:1]: I originally was trying to prevent taking pictures and pan/tilting from happening twice, but I think that your suggestion is better since readability and re-usability are probably more important on a demo procedure. I changed it so that IdentifySampleLocationDemo now reuses IdentifySampleTarget for both runs.

Comment (Samahu) [reply ow_plexil/src/plans/IdentifySampleLocationDemo.plp:1]: It is not a big deal to re-take pictures every time you need to analyze the site. Unless you want to conserve energy (by avoiding to move the antenna) but we can leave such optimizations to the end-users of autonomy to solve.

File: ow_plexil/scripts/identify_sample_location.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    in the image (usually shadows) and find the countours. The 2d center location is taken for these contours
+    and are sorted in descending order, we then draw/store the largest 5 contours/center points to return.'''
+    try:
+      image = self.bridge.imgmsg_to_cv2(location_image, "bgr8")
+    except CvBridgeError as err:
+      rospy.logerror("CV Bridge error: {0}".format(err))
+      return None, None, None
+
+    #defaults to Dark Spot if argument doesnt match or none is given
+    if filter_type == "Brown":
```
Comment (Samahu) [inline ow_plexil/scripts/identify_sample_location.py:135]: Optional, consider this:
```suggestion
    if filter_type.lower() == "brown":
```

File: ow_plexil/scripts/identify_sample_location.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+     #filtering brown spots on image
+      hsv = cv.cvtColor(image, cv.COLOR_BGR2HSV)
+      lower = np.array([0,16,32])
+      upper = np.array([19,255,255])
+      thresh = cv.inRange(hsv, lower, upper)
+    else:
+      #filtering dark spots on image
+      gray = cv.cvtColor(image, cv.COLOR_BGR2GRAY)
+      ret, thresh = cv.threshold(gray, 15, 255,cv.THRESH_BINARY_INV)
```
Comment (Samahu) [inline ow_plexil/scripts/identify_sample_location.py:144]: I think it better to be explicit about the selected filter rather defaulting to "black". 
```suggestion
    elif filter_type.lower() == "black":
      #filtering dark spots on image
      gray = cv.cvtColor(image, cv.COLOR_BGR2GRAY)
      ret, thresh = cv.threshold(gray, 15, 255,cv.THRESH_BINARY_INV)
    else:
       rospy.logerror("Unknown/Unsupported filter type specified!")   
```

File: ow_plexil/scripts/identify_sample_location.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+    #find contours and sort them largest to smallest
+    _, contours, hierarchy = cv.findContours(thresh, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
```
Comment (Samahu) [inline ow_plexil/scripts/identify_sample_location.py:147]:
```suggestion
    _, contours, _ = cv.findContours(thresh, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
```

File: ow_plexil/scripts/identify_sample_location.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    '''Given a list of 2d points we find the corresponding 3d point within the pointcloud. 
+    Takes the first valid point (largest contour with valid point) and transforms it into the
+    base_link frame. Returns the new 3d sample point, its corresponding contour and the original
+    2d location on the image.'''
+    if sample_points_2d:
+      site_coordinate = None
+      index = 0
+      #finds corresponding 3d point in point cloud from 2d pixel point takes the largest (first valid) point
+      for i in sensor_msgs.point_cloud2.read_points(point_cloud, uvs=sample_points_2d):
+        if(np.isnan(i[0]) or np.isnan(i[1]) or np.isnan([2])):
```
Comment (Samahu) [inline ow_plexil/scripts/identify_sample_location.py:178]: Outer parentheses are not needed, also note the third is isnan seems incorrect
```suggestion
        if np.isnan(i[0]) or np.isnan(i[1]) or np.isnan(i[2]):
```

**Review by Samahu — CHANGES_REQUESTED** [review]
Some comments needs your attention

File: ow_plexil/scripts/identify_sample_location.py (hunk 6 above)
Comment (Keeganfn) [reply ow_plexil/scripts/identify_sample_location.py]: Good catch, thanks!

**Review by Samahu — APPROVED** [review]
Can't spend more time on this .. I think it is in a good shape .. Good work!

File: ow_plexil/src/plans/ReferenceMission1.plp [hunk 7]
```diff
@@ -76,11 +77,11 @@ ReferenceMission1: Concurrence
     log_info ("Starting Reference Mission 1, Sol 0 ...");
     LibraryCall ImageLandingSite (InstanceName = "ReferenceMission1",
                                   IgnoreCrash = true);
-    LibraryCall Unstow();
```
Comment (kmdalal) [inline ow_plexil/src/plans/ReferenceMission1.plp:79]: Why was this removed?  Unstow is required before doing anything else with the arm.

File: ow_plexil/src/plans/ReferenceMission1.plp [hunk 8]
```diff
@@ -97,7 +98,6 @@ ReferenceMission1: Concurrence
     else
     {
       log_error ("Failed to find ground, aborting.");
-      LibraryCall Stow();
```
Comment (kmdalal) [inline ow_plexil/src/plans/ReferenceMission1.plp:100]: Why remove this?  Stowing the arm at completion is part of the normal conops (concept of operations).

File: ow_plexil/src/plexil-adapter/OwInterface.h [hunk 9]
```diff
@@ -209,10 +217,12 @@ class OwInterface
   std::unique_ptr<DigCircularActionClient> m_digCircularClient;
   std::unique_ptr<DigLinearActionClient> m_digLinearClient;
   std::unique_ptr<DeliverActionClient> m_deliverClient;
+  std::unique_ptr<IdentifySampleLocationActionClient> m_identifySampleLocationClient;
 
   // Antenna state - note that pan and tilt can be concurrent.
   double m_currentPan, m_currentTilt;
   double m_goalPan, m_goalTilt;      // commanded pan/tilt values
+  bool m_pointCloudRecieved;
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.h:225]: This should be initialized in the default constructor (even though in current usage it doesn't seem to matter.

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 10]
```diff
@@ -328,12 +333,32 @@ void OwInterface::managePanTilt (const string& opname,
 void OwInterface::cameraCallback (const sensor_msgs::Image::ConstPtr& msg)
 {
   // NOTE: the received image is ignored for now.
-
+  m_pointCloudRecieved = false;
   if (operationRunning (Op_TakePicture)) {
+    ros::Rate rate(10);
+    int timeout = 0;
+    // We wait for the pointcloud as well or the 5 sec timeout before marking as finished
+    while(m_pointCloudRecieved == false && timeout < 50){
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:341]: Best to make a named constant for the timeout (50) which can be defined local to this file.   I don't have a suggestion for a better approach than timeout -- this seems reasonable.

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 11]
```diff
@@ -328,12 +333,32 @@ void OwInterface::managePanTilt (const string& opname,
...
+        ros::spinOnce();
+        rate.sleep();
+        timeout+=1;
+    }
+    if(timeout == 50){
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:346]: Same comment as previous about using a constant.

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 12]
```diff
@@ -328,12 +333,32 @@ void OwInterface::managePanTilt (const string& opname,
...
   }
 }
 
+void OwInterface::pointCloudCallback (const sensor_msgs::PointCloud2::ConstPtr& msg)
+{
+  // NOTE: the received pointcloud is ignored for now.
+  if (operationRunning (Op_TakePicture)) {
+    //mark as recieved
+    m_pointCloudRecieved = true;
+    mark_operation_finished (Op_TakePicture, Running.at (Op_TakePicture));
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:359]: I don't think we want to mark the operation finished here, since that is done in the cameraCallback, which waits for this callback.

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 13]
```diff
@@ -424,6 +449,28 @@ static void guarded_move_done_cb
   publish ("GroundPosition", GroundPosition);
 }
 
+//Since template is outside the scope of the class we cannot use a member variable and 
+//instead have to use a global one like in the guarded_move_done_cb above.
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:453]: Yes, this is fine for now.  Technically the variables are static and not global (since they are not exported).

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 14]
```diff
@@ -424,6 +449,28 @@ static void guarded_move_done_cb
...
+static std::vector<double> m_sample_point;
+static bool got_sample_location_result = false;
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:455]: I'm not sure if we have a naming standard for static variables, but "m_" should be reserved for classes.  I'll suggest SamplePoint and GotSampleLocationResult to be consistent with the other static variables in this file/package.

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 15]
```diff
@@ -536,6 +587,7 @@ void OwInterface::initialize()
     m_digCircularClient = std::make_unique<DigCircularActionClient>(Op_DigCircular, true);
     m_digLinearClient = std::make_unique<DigLinearActionClient>(Op_DigLinear, true);
     m_deliverClient = std::make_unique<DeliverActionClient>(Op_Deliver, true);
+    m_identifySampleLocationClient = std::make_unique<IdentifySampleLocationActionClient>(Op_IdentifySampleLocation, true);
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:590]: Please shorten this line (by splitting it) for better readability -- 120 is max in our coding guidelines.  I like to keep lines as close to 80 as possible, though this is an old tradition :)

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 16]
```diff
@@ -804,6 +859,44 @@ void OwInterface::guardedMoveAction (double x, double y, double z,
     guarded_move_done_cb<GuardedMove, ow_lander::GuardedMoveResultConstPtr>);
 }
 
+std::vector<double> OwInterface::identifySampleLocation (int num_images, std::string filter_type, int id)
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:862]: const std::string& is better, and idiomatic.  

Style comment: for more readability and shorter lines, you could add a `using std::string;` at the top of the file, and just say `string` here and everywhere else in this file.

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 17]
```diff
@@ -804,6 +859,44 @@ void OwInterface::guardedMoveAction (double x, double y, double z,
...
+  got_sample_location_result = false;
+  if (! mark_operation_running (Op_IdentifySampleLocation, id)) return m_sample_point;
+  thread action_thread (&OwInterface::identifySampleLocationAction, this, num_images, filter_type, id);
+  action_thread.detach();
+
+  ros::Rate rate(10);
+  int timeout = 0;
+  /* We wait for a result from the action server or the 5 sec timeout before returning
+  I dont think a timeout is strictly necessary here but because this is blocking
+  I put it in for safety */
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:874]: This is our first Plexil command that returns a value, so there is no precedent for how to implement it.  Since the returned value is implemented as that of this function, this function must wait for it.  And since the only Plexil calls of this command are Synchronous, this is okay.   Ideally, the C++ implementation should not block, and thus support asynchronous value return in Plexil (this is allowed there...the assigned variable is UNKNOWN until the value arrives).  But in practice, most value-returning commands would be called synchronously.

In short, the wait/timeout here seems okay, and I don't have anything better to suggest.

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 18]
```diff
@@ -804,6 +859,44 @@ void OwInterface::guardedMoveAction (double x, double y, double z,
...
+  while(got_sample_location_result == false && timeout < 50){
+      ros::spinOnce();
+      rate.sleep();
+      timeout+=1;
+  }
+  if(timeout == 50){
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:880]: Same as earlier comment, use a named constant for 50.   Also, good to embed the unit in the name, e.g. SampleTimeoutSeconds.

**Comment by kmdalal** [comment]
I've finished reviewing and commenting on the code -- most were minor.   I also reformatted and slightly edited the comments in two of the PLEXIL files, and pushed them.    It's good to use punctuated sentences and filled paragraphs for such explanations.

The test appears to work for me, but I didn't see the sample target outlines today, and I already forgot to how to add the marker -- could you please add the instructions to plans/README.md along with an entry for the new test plan?

I'll retry the test after your final round of code edits.

File: ow_plexil/src/plans/ReferenceMission1.plp (hunk 7 above)
Comment (Keeganfn) [reply ow_plexil/src/plans/ReferenceMission1.plp:79]: I removed it because it was blocking the camera's view, but I forgot to add an unstow into the IdentifySampleTarget before the guarded move. I will make that fix!

File: ow_plexil/src/plans/ReferenceMission1.plp (hunk 8 above)
Comment (Keeganfn) [reply ow_plexil/src/plans/ReferenceMission1.plp:100]: Since the unstow is now only occurring if IdentifySampleTarget finds a valid point, I need to move the stow call to after all arm operations are finished (Which I forgot to do before).

File: ow_plexil/src/plexil-adapter/OwInterface.cpp (hunk 12 above)
Comment (Keeganfn) [reply ow_plexil/src/plexil-adapter/OwInterface.cpp:359]: Good catch, I must have left that in accidentally when copy pasting!

**Comment by kmdalal** [comment]
Satisfied with code - thanks for the updates.  I just pushed a very minor edit to OwInterface.h -- reindented a few lines, mostly not in your changes.

File: ow_plexil/src/plans/README.md [hunk 19]
```diff
@@ -60,3 +69,26 @@ The following represent the success criteria of the Reference Mission 2 plan:
 response to a fault
 	- Faults can be injected via rqt, a python script, or the command line
  ([Fault Injection Tutorial](https://github.com/nasa/ow_simulator/blob/master/ow_faults/README.md))
+
+
+### IdentifySampleLocationDemo ###
+
+The following represent the success criteria of the IdentifySampleLocationDemo plan: 
+- In the Rqt image viewer select the /sample_location topic. Once a point is selected
```
Comment (kmdalal) [inline ow_plexil/src/plans/README.md:77]: I cannot find this selection (see screenshot).  Should we be using the melodic-devel branch of ow_simulator?
rviz-topics

File: ow_plexil/src/plans/README.md [hunk 20]
```diff
@@ -60,3 +69,26 @@ The following represent the success criteria of the Reference Mission 2 plan:
...
+  the outlined contours of potential sample locations and the chosen target should show up.
+  The green dot represents the selected point.
+- In Rviz to vizualize the 3d point and to see if the guarded move went to the correct location:
+  - Select Add in the left menu under "Displays"
+  - Select "By topic"
+  - Scroll down and select the marker topic /sample_point_visualization and select ok.
```
Comment (kmdalal) [inline ow_plexil/src/plans/README.md:83]: I don't see /sample_point_visualization (see screenshot of menu).  I'm on melodic-devel of ow_simulator.
rviz-topics

Comment (Keeganfn) [reply ow_plexil/src/plans/README.md:83]: Hmm that is really odd, I am using the melodic-devel branch of ow-simulator. For me it is located at the very bottom if you scroll all the way down:

MarkerLocation

Is both the main ow launch file and the ow_plexil launch file running?

Comment (kmdalal) [reply ow_plexil/src/plans/README.md:83]: Never mind, I found it, after starting the Plexil node.   This was needed to see the rqt topic as well.  I'm editing the README a bit...

**Comment by kmdalal** [comment]
There was a spurious tilt error.  Isn't the end value within tolerance?
```
[ INFO]: Starting TiltAntenna: 50.000000 degrees (0.872665 radians)
[ERROR]: TiltAntenna timed out
[ERROR]: TiltAntenna failed. Ended at 50.659735 degrees, goal was 50.000000.
```

**Comment by Keeganfn** [comment]
> There was a spurious tilt error. Isn't the end value within tolerance?
> 
> ```
> [ INFO]: Starting TiltAntenna: 50.000000 degrees (0.872665 radians)
> [ERROR]: TiltAntenna timed out
> [ERROR]: TiltAntenna failed. Ended at 50.659735 degrees, goal was 50.000000.
> ```

Yeah that happens pretty reliably with me as well. The tolerance that it currently has is 0.4 both ways so its just outside of the tolerance range. Once the fledgling is approved I can make the action client changes to the pan and tilt which should fix all of those issues. I can also do it before the fledgling gets pushed if you would prefer that.

**Comment by kmdalal** [comment]
@Keeganfn  I see now that the tilt was slightly out of tolerance.  Feel free to add the pan/tilt action client (on melodic-devel) if you'd like to.  It's not  a priority for this work, or the fledgling.

**Comment by kmdalal** [comment]
This work requires the Plexil node (autonomy) to be started in order to add features to the simulator (rqt, Rviz) that it uses.   Could those features be moved to ow_simulator after this branch is merged?  Are they meaningful to have there?

**Comment by kmdalal** [comment]
I just pushed plans/README.md with some clarifications about seeing the new menu entries, and minor reformatting.

**Comment by kmdalal** [comment]
I tried this with the europa_terminator_workspace launch.  It found one sample point, and then failed to find another, and the plan does not terminate.   Full output:
```
[ INFO]: PLEXIL: Testing Dark filter in IdentifySampleLocation
[ INFO]: Starting TiltAntenna: 50.000000 degrees (0.872665 radians)
[ INFO]: Starting PanAntenna: 65.000000 degrees (1.134464 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Starting PanAntenna: 70.000000 degrees (1.221730 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Starting PanAntenna: 75.000000 degrees (1.308997 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Sending goal to action IdentifySampleLocation
[ INFO]: IdentifySampleLocation started...
[ INFO]: Possible sample location identified at (1.635319, -0.043864, -0.166163)
[ INFO]: Sending goal to action Unstow
[ INFO]: Unstow started...
[ INFO]: Unstow finished in state SUCCEEDED
[ INFO]: Sending goal to action GuardedMove
[ INFO]: GuardedMove started...
[ INFO]: GuardedMove finished in state SUCCEEDED
[ INFO]: PLEXIL: Testing Brown filter in IdentifySampleLocation
[ INFO]: Starting TiltAntenna: 50.000000 degrees (0.872665 radians)
[ERROR]: TiltAntenna timed out
[ERROR]: TiltAntenna failed. Ended at 50.639979 degrees, goal was 50.000000.
[ INFO]: Starting PanAntenna: 65.000000 degrees (1.134464 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Starting PanAntenna: 70.000000 degrees (1.221730 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Starting PanAntenna: 75.000000 degrees (1.308997 radians)
[ INFO]: Capturing stereo image using left image trigger.
[ INFO]: Sending goal to action IdentifySampleLocation
[ INFO]: IdentifySampleLocation started...
[INFO]: Could not find valid sample location
[ERROR]: Could not get sample point from IdentifySampleLocation
```

**Comment by Keeganfn** [comment]
@kmdalal Both of the visualization components are contained in the identify_sample_location.py file in the scripts directory of ow_plexil as publishers. I don't necessarily know whether that should be a lander operation or a autonomy operation, as it doesn't really fit completely in one or the other, but I'm open to ideas. My initial thought was since it directly interfaces with mostly autonomy operations that it is best suited for the ow_plexil package.

I also was able to reproduce the error you were talking about. Since it is all black and white on that workspace no point should be found most of the time for the "Brown" filter. It turns out that PLEXIL does not like being passed an empty Value vector and freezes up. I added a small fix that accounts for this and will now exit gracefully instead of locking up.

**Comment by kmdalal** [comment]
I think the image analysis support is a part of autonomy, but not logically a part of the ow_plexil package.  However, let's leave it be, and we can discuss a reorganization later.

Thanks for the fix, worked!  I'm satisfied with my testing.  I'm assuming that image analysis is deterministic, i.e. it will always return the same result for a given terrain and antenna angle; is this so?

File: ow_plexil/src/plexil-adapter/OwInterface.h [hunk 21]
```diff
@@ -127,18 +134,21 @@ class OwInterface
   void unstowAction (int id);
   void stowAction (int id);
   void grindAction (double x, double y, double depth, double length,
-               bool parallel, double ground_pos, int id);
+                    bool parallel, double ground_pos, int id);
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.h:137]: I just committed a fix that removes all the tab characters.

Comment (Keeganfn) [reply ow_plexil/src/plexil-adapter/OwInterface.h:137]: Thanks! I don't know how tab characters keep on sneaking in. I think my vimrc must be screwed up.

**Comment by Keeganfn** [comment]
> I think the image analysis support is a part of autonomy, but not logically a part of the ow_plexil package. However, let's leave it be, and we can discuss a reorganization later.
> 
> Thanks for the fix, worked! I'm satisfied with my testing. I'm assuming that image analysis is deterministic, i.e. it will always return the same result for a given terrain and antenna angle; is this so?

Yeah I believe it is for the most part. It also depends on the point cloud generation being consistent, but it should always be within a very small degree of tolerance.

**Review by kmdalal — APPROVED** [review]
I think this will be a very useful addition of illustrative autonomy to the system.  Great work!
